File: src/chessy/game/state.py
"""Game state machine — tracks phase transitions and move history."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from chessy.core.enums import Color, GameResult
from chessy.core.move_generator import MoveGenerator
from chessy.core.notation import STARTING_FEN, position_from_fen, position_to_fen
from chessy.core.rules import Rules
from chessy.game.interfaces import DrawOffer, GamePhase

if TYPE_CHECKING:
    from chessy.core.move import Move
    from chessy.core.position import Position


@dataclass
class MoveRecord:
    """A single entry in the move history."""

    move: Move
    san: str
    fen_after: str
    was_check: bool = False
    was_capture: bool = False


@dataclass
class GameState:
    """Manages game lifecycle: phase, result, move history, draw offers.

    This is a pure data/logic class — no threading, no UI.
    """

    position: Position = field(init=False)
    phase: GamePhase = field(default=GamePhase.NOT_STARTED, init=False)
    result: GameResult = field(default=GameResult.IN_PROGRESS, init=False)
    draw_offer: DrawOffer = field(default=DrawOffer.NONE, init=False)
    move_history: list[MoveRecord] = field(default_factory=list, init=False)
    start_fen: str = field(default=STARTING_FEN, init=False)

    # ── Initialisation ───────────────────────────────────────────────────

    def setup(self, fen: str | None = None) -> None:
        """Initialise (or reset) the game."""
        self.start_fen = fen or STARTING_FEN
        self.position = position_from_fen(self.start_fen)
        self.phase = GamePhase.AWAITING_MOVE
        self.result = GameResult.IN_PROGRESS
        self.draw_offer = DrawOffer.NONE
        self.move_history.clear()

    # ── Move application ─────────────────────────────────────────────────

    def apply_move(self, move: Move) -> MoveRecord:
        """Apply a validated move and return the history record.

        Caller is responsible for legality check.
        """
        from chessy.core.notation import move_to_san

        board = self.position.board
        was_capture = board[move.to_sq] is not None

        san = move_to_san(self.position, move)
        self.position.make_move(move)

        fen_after = position_to_fen(self.position)
        gen = MoveGenerator(self.position)
        was_check = gen.is_in_check(self.position.side_to_move)

        record = MoveRecord(
            move=move,
            san=san,
            fen_after=fen_after,
            was_check=was_check,
            was_capture=was_capture,
        )
        self.move_history.append(record)

        # Check for game-ending conditions
        self._check_game_over()
        self.draw_offer = DrawOffer.NONE  # any move cancels a pending offer

        return record

    def undo_last_move(self) -> Move | None:
        """Undo the last move. Returns the undone Move, or None if empty."""
        if not self.move_history:
            return None

        record = self.move_history.pop()
        self.position.unmake_move(record.move)

        # Reset result if we un-did a game-ending move
        if self.result != GameResult.IN_PROGRESS:
            self.result = GameResult.IN_PROGRESS
            self.phase = GamePhase.AWAITING_MOVE

        return record.move

    # ── Resignation / draw ───────────────────────────────────────────────

    def resign(self, color: Color) -> None:
        self.result = (
            GameResult.BLACK_WINS if color == Color.WHITE else GameResult.WHITE_WINS
        )
        self.phase = GamePhase.GAME_OVER

    def set_draw(self) -> None:
        self.result = GameResult.DRAW
        self.phase = GamePhase.GAME_OVER

    def flag_fall(self, color: Color) -> None:
        """Time ran out for *color*."""
        self.result = (
            GameResult.BLACK_WINS if color == Color.WHITE else GameResult.WHITE_WINS
        )
        self.phase = GamePhase.GAME_OVER

    # ── Query helpers ────────────────────────────────────────────────────

    @property
    def side_to_move(self) -> Color:
        return self.position.side_to_move

    @property
    def is_game_over(self) -> bool:
        return self.phase == GamePhase.GAME_OVER

    @property
    def ply_count(self) -> int:
        """Number of half-moves played."""
        return len(self.move_history)

    @property
    def fullmove_display(self) -> int:
        """Current full-move number for display."""
        return (self.ply_count // 2) + 1

    def legal_moves(self) -> list[Move]:
        """Legal moves in the current position."""
        gen = MoveGenerator(self.position)
        return gen.generate_legal_moves()

    # ── Internal ─────────────────────────────────────────────────────────

    def _check_game_over(self) -> None:
        result = Rules.game_result(self.position)
        if result != GameResult.IN_PROGRESS:
            self.result = result
            self.phase = GamePhase.GAME_OVER

File: src/game/state.cpp
#include <chessie/game/state.hpp>

#include <chessie/movegen.hpp>
#include <chessie/notation/san.hpp>

namespace chessie {

GameState::GameState() {
    position_ = Position::initial();
    start_fen = std::string(kStartingFen);
}

void GameState::setup(std::optional<std::string_view> fen) {
    start_fen = fen.has_value() ? std::string(*fen) : std::string(kStartingFen);
    position_ = fen.has_value() ? Position::from_fen(*fen) : Position::initial();
    phase = GamePhase::AwaitingMove;
    result = GameResult::InProgress;
    end_reason = GameEndReason::None;
    draw_offer = DrawOffer::None;
    draw_offer_by.reset();
    move_history.clear();
}

MoveRecord GameState::apply_move(Move move) {
    const Board& board = position_.board();
    const bool was_capture = board.piece_at(move.to_sq).type != PieceType::None ||
                             move.flag == MoveFlag::EnPassant;

    const std::string san = move_to_san(position_, move);
    position_.make_move(move);

    const std::string fen_after = position_.to_fen();
    const bool was_check = position_.is_in_check(position_.side_to_move());

    MoveRecord record{
        move,
        san,
        fen_after,
        was_check,
        was_capture,
    };
    move_history.push_back(record);

    check_game_over();
    draw_offer = DrawOffer::None;
    draw_offer_by.reset();

    return record;
}

std::optional<Move> GameState::undo_last_move() {
    if (move_history.empty()) {
        return std::nullopt;
    }

    const MoveRecord record = move_history.back();
    move_history.pop_back();
    position_.unmake_move(record.move);

    if (result != GameResult::InProgress) {
        result = GameResult::InProgress;
        end_reason = GameEndReason::None;
        phase = GamePhase::AwaitingMove;
    }

    return record.move;
}

void GameState::resign(Color color) {
    result = color == Color::White ? GameResult::BlackWins : GameResult::WhiteWins;
    end_reason = GameEndReason::Resign;
    phase = GamePhase::GameOver;
}

void GameState::set_draw(GameEndReason reason) {
    result = GameResult::Draw;
    end_reason = reason;
    phase = GamePhase::GameOver;
}

void GameState::flag_fall(Color color) {
    result = color == Color::White ? GameResult::BlackWins : GameResult::WhiteWins;
    end_reason = GameEndReason::FlagFall;
    phase = GamePhase::GameOver;
}

bool GameState::can_claim_draw_by_rule() const {
    return Rules::is_claimable_draw(position_);
}

bool GameState::claim_draw_by_rule() {
    if (is_game_over() || !can_claim_draw_by_rule()) {
        return false;
    }
    set_draw(GameEndReason::DrawRule);
    return true;
}

MoveList GameState::legal_moves() {
    return movegen::legal(position_);
}

void GameState::check_game_over() {
    const MoveList legal = movegen::legal(position_);

    if (legal.empty()) {
        if (position_.is_in_check(position_.side_to_move())) {
            result = position_.side_to_move() == Color::White ? GameResult::BlackWins
                                                                : GameResult::WhiteWins;
            end_reason = GameEndReason::Checkmate;
        } else {
            result = GameResult::Draw;
            end_reason = GameEndReason::Stalemate;
        }
        phase = GamePhase::GameOver;
        return;
    }

    if (Rules::is_automatic_draw(position_)) {
        result = GameResult::Draw;
        end_reason = GameEndReason::DrawRule;
        phase = GamePhase::GameOver;
    }
}

}  // namespace chessie
